iexec-sms-library/src/main/java/com/iexec/sms/api/SmsClientProvider.java

```diff
@@ -0,0 +1,89 @@
+ */
+@Slf4j
+public class SmsClientProvider {
+    // TODO: purge once task has been completed
+    private final Map<String, Optional<String>> taskIdToSmsUrl = new HashMap<>();
+    private final Map<String, SmsClient> urlToSmsClient = new HashMap<>();
+
+    private final IexecHubAbstractService iexecHubService;
+    private final Logger.Level loggerLevel;
+
+    public SmsClientProvider(IexecHubAbstractService iexecHubService) {
+        this.iexecHubService = iexecHubService;
+        this.loggerLevel = Logger.Level.NONE;
+    }
+
+    public SmsClientProvider(IexecHubAbstractService iexecHubService,
+                             Logger.Level loggerLevel) {
+        this.iexecHubService = iexecHubService;
+        this.loggerLevel = loggerLevel;
+    }
+
+
+    /**
+     * Retrieves the specified SMS URL for this task, then:
+     * <ul>
+     *     <li>If this SMS has already been accessed, returns the already-constructed {@link SmsClient};</li>
+     *     <li>Otherwise, constructs, stores and returns a new {@link SmsClient}.</li>
+     * </ul>
+     *
+     * @param chainTaskId ID of the task the specified SMS URL should be retrieved.
+     * @throws SmsClientCreationException if SMS URL can't be retrieved.
+     * @return An instance of {@link SmsClient} pointing on the task's specified SMS.
+     */
+    public SmsClient getOrCreateSmsClientForTask(String chainTaskId) {
+        final Optional<String> smsUrl = getSmsUrlForTask(chainTaskId);
+        if (smsUrl.isEmpty() || StringUtils.isEmpty(smsUrl.get())) {
+            throw new SmsClientCreationException("No SMS URL defined for given task [chainTaskId: " + chainTaskId +"]");
+        }
+
+        return urlToSmsClient.computeIfAbsent(smsUrl.get(), url -> SmsClientBuilder.getInstance(loggerLevel, url));
+    }
+
+    /**
+     * Retrieves the SMS URL defined for this task and caches the result.
+     * </p>
+     * If the task has already been initialized, then gets the URL from the {@link TaskDescription}.
+     * Otherwise, gets the URL from the {@link ChainDeal}.
+     *
+     * @param chainTaskId ID of the task the specified SMS URL should be retrieved.
+     * @return The SMS URL defined for this task.
+     */
+    Optional<String> getSmsUrlForTask(String chainTaskId) {
+        if (taskIdToSmsUrl.containsKey(chainTaskId)) {
+            return taskIdToSmsUrl.get(chainTaskId);
+        }
+
+        final Optional<String> smsUrl;
+
+        final TaskDescription taskDescription = iexecHubService.getTaskDescription(chainTaskId);
+        if (taskDescription != null) {
+            smsUrl = Optional.ofNullable(taskDescription.getSmsUrl());
+        } else {
+            // Fallback: if task is not initialized yet,
+            // we can still get its SMS url in its deal.
+            final Optional<ChainDeal> chainDeal = iexecHubService.getChainDeal(chainTaskId);
+            smsUrl = chainDeal
+                    .map(deal -> deal.getParams().getIexecSmsUrl());
```

Comment:
Why don't we simply do this single call in all cases ?
Blockchain calls are costly and it would be best to guarantee that data can be retrieved in one call.

Reply:
Not 100% sure, but I'd guess retrieving only the task description when possible would be faster than getting the whole deal. Any clue on that @jeremyjams?

Reply:
Well, if you read the code, you will see getChainDeal is already called from getTaskdescription in iexecHubService.
I still think this code could be simplified, one way or another:
https://github.com/iExecBlockchainComputing/iexec-common/blob/6b236e8246ac7ff63f4c4d8b83da63b16755c8ce/src/main/java/com/iexec/common/chain/IexecHubAbstractService.java#L1166

Reply:
https://github.com/iExecBlockchainComputing/iexec-common/blob/6b236e8246ac7ff63f4c4d8b83da63b16755c8ce/src/main/java/com/iexec/common/chain/IexecHubAbstractService.java#L1153

Reply:
Best option is to only call `taskDescription.getSmsUrl()` once yes.
I was thinking adding sms url validation inside `getTaskDescription` for tasks having TEE type (but we can do that later depending on next monday meeting)

Reply:
true, we will avoid a nice bug with this discussion

Reply:
Here is a new version where I have split both versions: https://github.com/iExecBlockchainComputing/iexec-sms/pull/116/commits/2e95031feadb283f80ffbc4356f1ea78cf4db9f0.
So that uninitialized tasks will retrieve their SMS URL from the `ChainDeal` while initialized tasks will retrieve them from their `TaskDescription`.